docs/protocol/extension_generate.md

```diff
@@ -0,0 +1,126 @@
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+-->
+
+# Generate Extension
+
+This document describes Triton's generate extension. The generate
+extension provides a simple text oriented endpoint schema for interacting
+with LLM models. This generate endpoint is specific to HTTP/REST frontend.
+
+## HTTP/REST
+
+In all JSON schemas shown in this document `$number`, `$string`, `$boolean`,
+`$object` and `$array` refer to the fundamental JSON types. #optional
+indicates an optional JSON field.
+
+Triton exposes the generate endpoint at the following URLs. The client may use
+HTTP POST request to different URLs for different response behavior, the
+endpoint will return the generate results on success or an error in the case of
+failure.
+
+```
+POST v2/models/${MODEL_NAME}[/versions/${MODEL_VERSION}]/generate
+
+POST v2/models/${MODEL_NAME}[/versions/${MODEL_VERSION}]/generate_stream
+```
+
+### generate v.s. generate_stream
+
+Both URLs expect the same request JSON object, and generate the same response
+JSON object. However, `generate` return exactly 1 response JSON object, while
+`generate_stream` may return various number of responses based on the inference
+results. `generate_stream` returns the responses as
+[Server-Sent Events](https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events)
+(SSE), where each response will be a "data" chunk in the HTTP response body.
+Also note that an error may be returned during inference, whereas the HTTP
+response code has been set at first response of the SSE, which can result in
+receiving error object while status code is success (200). Therefore the user
+must always check whether an error object is received when generating responses
+through `generate_stream`.
+
+### Generate Request JSON Object
+
+The generate request object, identified as *$generate_request*, is
+required in the HTTP body of the POST request. The model name and
+(optionally) version must be available in the URL. If a version is not
+provided the server may choose a version based on its own policies or
+return an error.
+
+    $generate_request =
+    {
+      "text_input" : $string,
+      "parameters" : $parameters #optional
+    }
```

Comment:
Do you want this to be part of the generate schema? Seems to be de-valuing the purpose of adding reserved "parameters" field.

Reply:
I think the parameters route might be easier for users, or at least I would find it easier as a user.

I think there could be a way to document both approaches without confusing users, if done right. This approach is good because it allows people to be more explicit with which are parameters, so the feature still has value (e.g. if an input and parameter have identical names). I think the question though should be less about what de-values a feature and more about what provides the most ease and value to users.

Reply:
I'm good either way - we can add a note:

```suggestion
   > [!NOTE] > Any additional properties in the request object are passed either as parameters or tensors based on model specification. 
```

but in all examples let's use 'parameters' for consistency (i.e. not show both forms). 

I believe we'll refine this further and may either lock the schema further or make it more flexible. Since this is provisional we won't guarantee backwards compat (yet)